### DISCUSSION TITLE
FieldHasher for other hash functions

### DISCUSSION THREAD

**Discussion by sid-alluri** [Consensys/gnark#824] [Q&A]
Hi, I am learning gnark and I was trying to use Poseidon instead of MiMC, which is present in most of the examples. However, I couldn't understand how I could use other hashes inside a circuit. Is there a way to convert a BinaryHasher to a FieldHasher?

**Comment by ivokub** [comment]
Hi, there difference between `BinaryHasher` and `FieldHasher` is essentially if the hash function reads in field elements (think of them as integers below some bound) or bytes. Conventionally, the hash functions took as inputs bytes and output some fixed-width byte digests (e.h. MD5, SHA1, SHA2, Blake2, SHA3 etc. etc), but in SNARK circuits the primitive types we're working on are field elements. It is actually quite expensive computationally for the prover to work on bytes (e.g. AND/XOR/OR operations etc.) and so recently there is a line of work to define "SNARK-friendly hash functions" which work on SNARK-friendly primitive types (finite field elements). This also requires changes to the hash function interface to take as inputs field elements (instead of bytes).

So, there is a principal difference between conventional hash functions and SNARK-friendly hash functions and there is not a convenient way to transfer one to another (it can be done by defining a 1-1 mapping between bytes and field elements). So I would say there isn't a way to convert BinaryHasher to FieldHasher trivially.

But Poseidon hash function is in the family of SNARK-friendly hash functions, so it should by default implement FieldHasher interface anyway. Do you have a particular implementation of Poseidion hash you're interested in?

**Comment by sid-alluri** [comment]
Thanks for the detailed explanation. I am able to run my code.

**Reply by jeljelih** [reply]
@sid-alluri Did you succeed in using Poseidon instead of MiMC in gnark? If you did, would you share your implementation?

**Reply by sid-alluri** [reply]
Hi, No, I haven't used Poseidon.
